### DISCUSSION TITLE
Difference between saveRelationshipsUsing and saveRelationshipsBeforeChildrenUsing

### DISCUSSION THREAD

**Discussion by demobiel** [filamentphp/filament#8141] [Help] [answered]
I can't get my head around those... not really sure what the difference is or when which one should be used...

**Comment by danharrin** [comment]
What are you trying to do

**Reply by demobiel** [reply]
hi Dan

I am trying to get a general understanding when to use which function. I am even not sure that those functions are meant for doing what I want to do. But as I don't see how and when they are triggered, or what their purpose is, I am a little lost / confused.

Looking for a way to have control on how related data is created / deleted, as I have the feeling I am making things way to complex.

For example: 

**Don't create related record in a HasOne relationship if related attribute not filled in, or delete it when it is emptied**

I have a customer - domain relationship where each customer can have one domain, with following relationship defined:
```
//Customer Model class
public function domain(): MorphOne
{
     return $this->morphOne(Domain::class, 'domainable');
}
```

and following form field in my CustomerResource file
```
 Forms\Components\Group::make()
    ->relationship('domain')
    ->schema([
        Forms\Components\TextInput::make('domain')
           ->unique(ignorable: fn($record) => $record)
    ])
```
The idea is that the user can enter a domain for his project, but when it is not filled in, it should not create a related domain record.

Visually this looks like this, where "domain" comes from a related table "domains".

<img width="629" alt="image" src="https://github.com/filamentphp/filament/assets/390530/ba81d0ac-de22-44d4-986e-187d83482d33">

By default Filament will try to create a domain record with domain = null on save of the customer resource if the domain field is not filled in, causing a mysql error as domain cannot be null.

In order to prevent this I use the saveRelationshipsBeforeChildrenUsing, to manage the update of the related domain  record (create/update/delete) (this code works btw)

This results in:
```
 Forms\Components\Group::make()
    ->relationship('domain')
    ->schema([
        Forms\Components\TextInput::make('domain')
           ->unique(ignorable: fn($record) => $record)
           ->regex('/^(?:[a-z0-9](?:[a-z0-9-æøå]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$/isu')
    ])->saveRelationshipsBeforeChildrenUsing(function (Customer $customer, Forms\Components\Component $component) {
          $state = $component->getState();
          if ($state['domain'] === '') {
                $customer->domain()->delete();
          } else {
                if ($customer->domain()->first()) {
                     $customer->domain()->update([
                          'domain' => $state['domain']
                      ]);
               } else {
                     $customer->domain()->create($state);
               }
           }
    }),

```

But it also semi works if I use ->saveRelationshipsUsing()... and I am not getting my head around which one to use, and I have the feeling I am making things way to complex.

if you could point me to the files where those are called during the lifecycle, that would help me a lot.

**Reply by demobiel** [reply]
hi Dan,

 I think I understood now! 
BeforeChildren -> is called before all childcomponents save their relationships, I also found another solution to my problem using the condition parameter of the relationship method.

Does this make sense (the goal is : delete the Domain record if the "domain" column is not filled in. I couldn't find any docs about this extra parameter, but reading the code it seems it will delete the related record.

````
 Forms\Components\Group::make()
         ->relationship(
                name: 'domain',
                condition: fn(Forms\Components\Component $component) => $component->getState()['domain'] != '')
          ->schema([
                Forms\Components\TextInput::make('domain')
                    ->unique(ignorable: fn($record) => $record)
                    ->regex('/^(?:[a-z0-9](?:[a-z0-9-æøå]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$/isu')
          ]),
````

**Reply by danharrin** [reply, marked as answer]
The condition parameter sounds right, yeah 👍
